### PR TITLE
pycc distutils integration

### PR DESCRIPTION
**Pull request by pitrou**
Based on PR #1476. This also includes documentation and API cleanups.

### CODE REVIEW

**Comment by seibert** [comment]
Is this the PR where it makes sense to add the user documentation describing how to do ahead of time compilation in their Python project?

**Comment by pitrou** [comment]
Yes, probably.

**Comment by seibert** [comment]
Time for a remerge now that #1476 has been merged to master.

**Comment by seibert** [comment]
On OS X 10.11, I'm seeing linker errors when I try to import a compiled module that uses NRT functions:
``` python
# src_mode.py
import numpy as np
from numba.pycc import CC

cc = CC('mode')

@cc.export('simple_mode', 'int64(int64[:])')
def simple_mode(x):
    sorted_x = np.sort(x)

    return sorted_x[0]

if __name__ == "__main__":
    cc.compile()
```
(Ignore the names.  This is cut down from a larger test.)

When I compile the module by running `python src_mode.py` and and then attempt to import the generated module, I get this error:
```
(numba_aot)Rover:mode stan$ python
Python 3.4.3 |Continuum Analytics, Inc.| (default, Mar  6 2015, 12:07:41)
[GCC 4.2.1 (Apple Inc. build 5577)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import mode
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: dlopen(/work/continuum/aot_demo/mode/mode.so, 2): Symbol not found: _NRT_MemInfo_alloc_safe_aligned
  Referenced from: /work/continuum/aot_demo/mode/mode.so
  Expected in: flat namespace
 in /work/continuum/aot_demo/mode/mode.so
```

**Comment by pitrou** [comment]
> On OS X 10.11, I'm seeing linker errors when I try to import a compiled module that uses NRT functions:

Hmm, I'm not sure what to say. Is the failure only with this PR or also on master? Also, is it only for NRT functionality? What happens if you use a helperlib function e.g. math.gamma()?

Adding `cc.verbose = True` helps you see what steps are performed.

**Comment by pitrou** [comment]
Oh, and I don't know what kind of toolchain is available on OS X, but you can try `nm  --defined-only name_of_the_so_file` to see what symbols are defined inside it.

**Comment by seibert** [comment]
I can make that error go away if I remove the `VISIBILITY_HIDDEN` before `NRT_MemInfo_alloc_safe_aligned` in nrt.h.  Then I get a similar linker error for another function declared in nrt.h

If I remove `VISIBILITY_HIDDEN` from 3 specific functions used by this code, then I get to linker error with a fourth function: `NRT_adapt_ndarray_from_python`.  I'm not sure what is deciding the visibility of this function.

Basically, it looks like the linker is failing to statically link NRT function calls into the extension on OS X for some reason.

**Comment by seibert** [comment]
OK, mystery solved (sort of).  I discovered that I had gcc 4.9 built by MacPorts in my path.  Once I removed that, reverting back to the system gcc (which is aliased to clang on OS X now), everything worked fine.  I have no idea why the MacPorts version of gcc was not statically linking as expected.

**Comment by sklam** [comment]
The problem also occur on gcc4.9 on ubuntu 14.04:

```
$ CC=gcc-4.9 python -m numba.tests.test_pycc_tresult TestCC.test_compile_helperlib -v
test_compile_helperlib (__main__.TestCC) ... C compiler: gcc-4.9 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC

creating /tmp/pycc-build-pycc_test_helperlib-jacj3i82/home
creating /tmp/pycc-build-pycc_test_helperlib-jacj3i82/home/sklam
creating /tmp/pycc-build-pycc_test_helperlib-jacj3i82/home/sklam/dev
creating /tmp/pycc-build-pycc_test_helperlib-jacj3i82/home/sklam/dev/numba
creating /tmp/pycc-build-pycc_test_helperlib-jacj3i82/home/sklam/dev/numba/numba
creating /tmp/pycc-build-pycc_test_helperlib-jacj3i82/home/sklam/dev/numba/numba/pycc
compile options: '-flto -DPYCC_MODULE_NAME=pycc_test_helperlib -DPYCC_USE_NRT=0 -I/home/sklam/miniconda/envs/numba_devel34/include/python3.4m -I/home/sklam/miniconda/envs/numba_devel34/lib/python3.4/site-packages/numpy/core/include -c'
Warning: Can't read registry to find the necessary compiler setting
Make sure that Python modules winreg, win32api or win32con are installed.
gcc-4.9: /home/sklam/dev/numba/numba/pycc/modulemixin.c
gcc-4.9: /home/sklam/dev/numba/numba/pycc/../_math_c99.c
gcc -pthread -shared -L/home/sklam/miniconda/envs/numba_devel34/lib -Wl,-rpath=/home/sklam/miniconda/envs/numba_devel34/lib,--no-as-needed -flto /tmp/pycc-build-pycc_test_helperlib-jacj3i82/pycc_test_helperlib.cpython-34m.o /tmp/pycc-build-pycc_test_helperlib-jacj3i82/home/sklam/dev/numba/numba/pycc/modulemixin.o /tmp/pycc-build-pycc_test_helperlib-jacj3i82/home/sklam/dev/numba/numba/_math_c99.o -L/home/sklam/miniconda/envs/numba_devel34/lib -L/home/sklam/miniconda/envs/numba_devel34/lib/python3.4/site-packages/numpy/core/lib -lpython3.4m -lnpymath -lm -o /tmp/test_pycc/pycc_test_helperlib.cpython-34m.so
lto1: fatal error: LTO_tags out of range: Range is 0 to 368, value is 756
compilation terminated.
lto-wrapper: gcc returned 1 exit status
/usr/bin/ld: lto-wrapper failed
collect2: error: ld returned 1 exit status
lto1: fatal error: LTO_tags out of range: Range is 0 to 368, value is 756
compilation terminated.
lto-wrapper: gcc returned 1 exit status
/usr/bin/ld: lto-wrapper failed
collect2: error: ld returned 1 exit status
ERROR

======================================================================
ERROR: test_compile_helperlib (__main__.TestCC)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/sklam/miniconda/envs/numba_devel34/lib/python3.4/distutils/unixccompiler.py", line 194, in link
    self.spawn(linker + ld_args)
  File "/home/sklam/miniconda/envs/numba_devel34/lib/python3.4/site-packages/numpy/distutils/ccompiler.py", line 28, in <lambda>
    m = lambda self, *args, **kw: func(self, *args, **kw)
  File "/home/sklam/miniconda/envs/numba_devel34/lib/python3.4/site-packages/numpy/distutils/ccompiler.py", line 72, in CCompiler_spawn
    raise DistutilsExecError('Command "%s" failed with exit status %d%s' % (cmd, s, msg))
distutils.errors.DistutilsExecError: Command "gcc -pthread -shared -L/home/sklam/miniconda/envs/numba_devel34/lib -Wl,-rpath=/home/sklam/miniconda/envs/numba_devel34/lib,--no-as-needed -flto /tmp/pycc-build-pycc_test_helperlib-jacj3i82/pycc_test_helperlib.cpython-34m.o /tmp/pycc-build-pycc_test_helperlib-jacj3i82/home/sklam/dev/numba/numba/pycc/modulemixin.o /tmp/pycc-build-pycc_test_helperlib-jacj3i82/home/sklam/dev/numba/numba/_math_c99.o -L/home/sklam/miniconda/envs/numba_devel34/lib -L/home/sklam/miniconda/envs/numba_devel34/lib/python3.4/site-packages/numpy/core/lib -lpython3.4m -lnpymath -lm -o /tmp/test_pycc/pycc_test_helperlib.cpython-34m.so" failed with exit status 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/sklam/dev/numba/numba/tests/test_pycc_tresult.py", line 184, in test_compile_helperlib
    with self.check_cc_compiled(self._test_module.cc_helperlib) as lib:
  File "/home/sklam/miniconda/envs/numba_devel34/lib/python3.4/contextlib.py", line 59, in __enter__
    return next(self.gen)
  File "/home/sklam/dev/numba/numba/tests/test_pycc_tresult.py", line 142, in check_cc_compiled
    cc.compile()
  File "/home/sklam/dev/numba/numba/pycc/cc.py", line 166, in compile
    extra_ldflags=extra_ldflags)
  File "/home/sklam/dev/numba/numba/pycc/platform.py", line 100, in link_shared
    extra_preargs=extra_ldflags)
  File "/home/sklam/miniconda/envs/numba_devel34/lib/python3.4/distutils/unixccompiler.py", line 196, in link
    raise LinkError(msg)
distutils.errors.LinkError: Command "gcc -pthread -shared -L/home/sklam/miniconda/envs/numba_devel34/lib -Wl,-rpath=/home/sklam/miniconda/envs/numba_devel34/lib,--no-as-needed -flto /tmp/pycc-build-pycc_test_helperlib-jacj3i82/pycc_test_helperlib.cpython-34m.o /tmp/pycc-build-pycc_test_helperlib-jacj3i82/home/sklam/dev/numba/numba/pycc/modulemixin.o /tmp/pycc-build-pycc_test_helperlib-jacj3i82/home/sklam/dev/numba/numba/_math_c99.o -L/home/sklam/miniconda/envs/numba_devel34/lib -L/home/sklam/miniconda/envs/numba_devel34/lib/python3.4/site-packages/numpy/core/lib -lpython3.4m -lnpymath -lm -o /tmp/test_pycc/pycc_test_helperlib.cpython-34m.so" failed with exit status 1

----------------------------------------------------------------------
Ran 1 test in 0.830s

FAILED (errors=1)
```

The problem appears to be related to: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=753134

**Comment by seibert** [comment]
Maybe for this PR we could disable the `-flto` flag and revisit this in the future when we have more information about where `-flto` works in practice.

**Comment by sklam** [comment]
The same problem occurred on gcc-5 too.

**Comment by pitrou** [comment]
Ok, it looks like `-flto` may still be quite buggy... (surprising, as it works fine for me with gcc 4.9.2 on Ubuntu 15.04... perhaps they backported additional patches).
I'll file a separate PR to disable it.
